modules/microshift-olm-deploy-op-disconnected.adoc

```diff
@@ -0,0 +1,24 @@
+//Module included in the following assemblies:
+//
+//* microshift_running_apps/microshift-operators-olm.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="microshift-adding-OLM-Operators-to-disconnected-cluster_{context}"]
+= Adding OLM-based Operators to a disconnected cluster
+
+For {microshift-short} clusters that are installed on restricted networks, also known as _disconnected clusters_, Operator Lifecycle Manager (OLM) by default cannot access sources hosted on remote registries because those remote sources require full internet connectivity.
+
+You can enable your cluster to use OLM-based Operators in a restricted network. The following steps are required to use OLM-based Operators in restricted networks:
+
+* Include OLM in your container image list for your mirror registry. See xref:../microshift_install/microshift-deploy-with-mirror-registry.adoc#microshift-get-mirror-reg-container-image-list_microshift-deploy-with-mirror-registry[Getting the mirror registry container image list].
+//Q:Or are we referring to link:https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html#installation-mirror-repository_installing-mirroring-installation-images[Mirroring images for a disconnected installation] but swapping out for oc-mirror?
+//or adding a different link such as: link:https://docs.openshift.com/container-platform/4.14/installing/disconnected_install/installing-mirroring-installation-images.html#olm-mirror-catalog_installing-mirroring-installation-images[Mirroring Operator catalogs for use with disconnected clusters]
+* Configure the system to use the mirror by updating your CRI-O configuration directly. `ImageContentSourcePolicy` is not supported in {microshift-short}. See link:https://access.redhat.com/documentation/en-us/red_hat_build_of_microshift/{product-version}/html/installing/microshift-deploy-with-mirror-registry.adoc#microshift-configuring-hosts-for-mirror_microshift-deployment-mirror[Configuring hosts for mirror registry access] for details.
+* Add a `CatalogSource` to the cluster so that the OLM catalog Operator can use the local catalog on the mirror registry.
+* Ensure that {microshift-short} is installed to run in the restricted network. See xref:../microshift_install/microshift-embed-in-rpm-ostree-offline-use.adoc#microshift-embed-in-rpm-ostree-offline-use[Embedding in a {op-system-ostree} image for offline use]
```

Comment:
"Embedding images" means that the ISO contains all ctr images required for running. Restricted network doesn't necessarily equals "offline mode".

It appears we actually have 3 modes of operations:
1. Full connectivity to the Internet (no need to mirror images)
2. Restricted/disconnected/air-gapped - MicroShift has access to some internal network, but not to the Internet - images must be mirrored to internal registry
3. Offline mode - MicroShift has no connectivity at all (or it's only intermittent) meaning images must be embedded into the osbuild blueprint

So, for 3rd option, we need a method to get all the ctr image refs for embedding, but not necessarily mirror them to a registry mirror because with image pull policy `IfNotPresent` they should be used from disk.

Reply:
I'm wondering how should we approach option 3.
I came up with these two approaches, but none seems ideal.

Approach 1: use side effect of `oc mirror` to get image refs to mirror. For example, this is what I got for cataloghub.io's cert-manager:
```
quay.io/jetstack/cert-manager-cainjector@sha256:ac5154525f99bd0872671613741aac1b7dcb9c0df988571a7618155ddb6fabd2=registry.example.com/oc-mirror/jetstack/cert-manager-cainjector:95fb9349
quay.io/jetstack/cert-manager-controller@sha256:2121d4250f5734ee097df243507d06536fc264140dba3425045a825ef597c79d=registry.example.com/oc-mirror/jetstack/cert-manager-controller:df9bd9f2
quay.io/jetstack/cert-manager-webhook@sha256:f45b21f770bf4676c732f19e2ef17c34f46ac75873a5e0aa25703d808b2e5566=registry.example.com/oc-mirror/jetstack/cert-manager-webhook:1fb77ce2
quay.io/operatorhubio/catalog:latest=registry.example.com/oc-mirror/operatorhubio/catalog:latest
quay.io/operatorhubio/cert-manager@sha256:2bdd113c638f3ed2b3f5beeb53c3f12e9d8e460bcc5618a9cfacd855dcd89afb=registry.example.com/oc-mirror/operatorhubio/cert-manager:e489fea8
```
though I'm not sure how it'll be handled in practice since it went from tags to SHAs.

Second approach is to inspect the operator in the catalog itself. For example, when I look at `configs/cert-manager/catalog.yaml` from `quay.io/operatorhubio/catalog:latest` image, there is following information:
```
image: quay.io/operatorhubio/cert-manager@sha256:2bdd113c638f3ed2b3f5beeb53c3f12e9d8e460bcc5618a9cfacd855dcd89afb
name: cert-manager.v1.13.3                                                                                                                                                                                                                                                                   
package: cert-manager   
...
relatedImages:                                                                                                     
- image: quay.io/jetstack/cert-manager-cainjector:v1.13.3                                                          
  name: ""                                                                                                                                                                                                                                                                                   
- image: quay.io/jetstack/cert-manager-controller:v1.13.3                                                          
  name: ""                                                                                                         
- image: quay.io/jetstack/cert-manager-webhook:v1.13.3                                                                                                                                                                                                                                       
  name: ""                                                                                                         
- image: quay.io/operatorhubio/cert-manager@sha256:2bdd113c638f3ed2b3f5beeb53c3f12e9d8e460bcc5618a9cfacd855dcd89afb
  name: ""
```
but this is too manual...

I wonder if there's something from OLM related tools to get just the refs for embedding

Reply:
Understood. Trying to make our write-up somewhat consistent with what we have in our Install docs. Which is: install by RPM in RHEL; embed in RHEL Edge with network connection (I assume always-on, or sometimes on); RHEL Edge for completely offline. I am not clear on which steps are the same for all three (making the catalog source object and Subscription CR, I imagine), and which steps we need to make different (and what those are for MicroShift).

Reply:
@pmtk @ShaunaDiaz 
for 3 (offline mode), the cluster has no any connectivity. is there internal registry for this cluster?
if there is no internal registry, where is the images pull when installing catalog and then installing operator?

I am sorry I can not understand how to install operator for such cluster if it has no internal registry.

Reply:
@pmtk 
thanks for the information. 

support you already get index image (for example, quay.io/operatorhubio/catalog:latest), bundle image and operator image list and add them into blueprint yaml file.
and then you build them to get iso file, and then create microshift cluster with this iso with offline mode.

when you create catalog source, what is the image url address? if it is still quay.io/operatorhubio/catalog:latest, olm should try to pull it from quay.io, not local container-storage?

suppose you create catalog source, when you install operator, olm-catalog will pull bundle image which is from index image (suppose quay.io/operatorhubio/cert-manager@sha256:32c3f3e40f7fe42f79be2b313d8d61c7849dba7fd2cf5d09ea5b5df99d45f6a3), and then olm will pull operator images which is from csv file (suppose quay.io/jetstack/cert-manager-webhook:v1.10.0), it also should try to pull them from quay.io, not local container-storage?

if it is the way, it should not work. (@kevinrizza please correct me if I am wrong, thanks)

so, maybe need to configure crio to mirror quay.io to others, what is it if there is no internal registry?

frankly I still can not understand how it works to install catalog and operator. maybe I can wait the detailed procedure to understand it.

Thanks

Reply:
OLM is specifically and purposefully unaware of any disconnected mirroring process. In a non microshift OCP case, oc-mirror generates the following:

- The json fbc file that defines the catalog metadata. That catalog metadata doesn't have anything special about it, all of its references still point to the normal image paths for the bundle images and related operator images. From that, you generate a catalog image using that json file as its source.
- A mapping.txt file that defines the set of images you need to configure cri-o to swap. That list of images is the catalog image itself that was built in the previous step plus all of the bundle images for each operator in the catalog and the list of related images that each bundle references.

For OCP, we take that mapping file and generate an image content source policy that users apply to the cluster. For microshift, I think the only difference in this whole process is that instead of using that mapping file to generate an ICSP, they'll need to take that mapping and configure cri-o themselves manually, in the same way we expect users to configure the rest of their images manually with microshift. I don't think there's any other complexity here other than that one step.

Reply:
I created offline ISO with OLM and cert-manager from OperatorHub.io and it works. The only caveat is that [if catalog's image is not using SHA, it will have pull policy `Always`](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/lib/image/image.go#L9-L17) and pod will not start. If the catalog source image uses SHA, then there's no problem.

I haven't had this problem with cert-manager which uses tags (:v1.13.3) because it has pull policy `IfNotPresent`

```
$ ip a | head -n 20
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet 10.44.0.1/32 scope global lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: ovs-system: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether f2:7a:80:5e:b7:33 brd ff:ff:ff:ff:ff:ff
3: br-ex: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 0a:59:00:00:00:01 brd ff:ff:ff:ff:ff:ff
    inet 10.44.0.0/32 scope global br-ex
       valid_lft forever preferred_lft forever
    inet 169.254.169.2/29 brd 169.254.169.7 scope global br-ex
       valid_lft forever preferred_lft forever
4: br-int: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 46:b9:ed:62:47:07 brd ff:ff:ff:ff:ff:ff
5: ovn-k8s-mp0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 2e:de:43:13:03:9f brd ff:ff:ff:ff:ff:ff
...
```

```
$ oc get pods -A
NAMESPACE                              NAME                                                              READY   STATUS      RESTARTS      AGE
kube-system                            csi-snapshot-controller-57dcf58b44-5kqqd                          1/1     Running     0             46m
kube-system                            csi-snapshot-webhook-867797d9bb-rcsks                             1/1     Running     0             46m
openshift-dns                          dns-default-c7srs                                                 2/2     Running     0             46m
openshift-dns                          node-resolver-jtqbq                                               1/1     Running     0             46m
openshift-ingress                      router-default-6d77c657d-55sw9                                    1/1     Running     0             46m
openshift-operator-lifecycle-manager   catalog-operator-68f6c78675-nt8rj                                 1/1     Running     0             46m
openshift-operator-lifecycle-manager   olm-operator-78f56fd657-55r92                                     1/1     Running     0             46m
openshift-operators                    0b9f8e8106e6bc92a5b3edb6791ceaab0e8a22f5493895798082899af7pp5k4   0/1     Completed   0             19m
openshift-operators                    cert-manager-7df8994ddb-qnxjw                                     1/1     Running     0             19m
openshift-operators                    cert-manager-cainjector-5746db8fd7-74njj                          1/1     Running     0             19m
openshift-operators                    cert-manager-webhook-84d67cfcb6-db7f6                             1/1     Running     0             19m
openshift-operators                    operatorhubio-catalog-5w47w                                       1/1     Running     0             19m
openshift-ovn-kubernetes               ovnkube-master-xjcxx                                              4/4     Running     1 (46m ago)   46m
openshift-ovn-kubernetes               ovnkube-node-sxc9s                                                1/1     Running     1 (46m ago)   46m
openshift-service-ca                   service-ca-6dc459dcf8-gn4xw                                       1/1     Running     0             46m
openshift-storage                      topolvm-controller-86cc645ff8-ndvd5                               5/5     Running     0             46m
openshift-storage                      topolvm-node-drx4l                                                4/4     Running     0             46m
```